src/MeMoMetabolite.py

```diff
@@ -280,30 +310,76 @@ def set_pKb(self, new_pKb:dict) -> None:
+            raise ValueError('''Entries length differs for new annotation and source dictionary. Consider to provide a source dictionary via the source value''')
+
+        # check if there is an old annotation which will be overwritten
         if self.annotations != {} and self.annotations != None:
+            warntrigger = True
+            old_annotations = deepcopy(self.annotations)
+        else:
+            warntrigger = False
+        
+        # delete the old information
+        self.annotations = dict()
+        self.annotations_source = dict()
+        
+        # sort both new dictionaries and remove duplicates 
+        for x in new_annotations.keys():
+            new_annolst = []
+            new_sourcelst = []
+            for anno,src in zip(new_annotations[x], source[x]):
+                if not anno in new_annolst:
+                    new_annolst.append(anno)
+                    new_sourcelst.append(src)
+            new_annolst,new_sourcelst = zip(*sorted(zip(new_annolst,new_sourcelst)))
+            self.annotations[x] = list(new_annolst)
+            self.annotations_source[x] = list(new_sourcelst)
+        
+        # trigger the warning and give some information to the user
+        if warntrigger == True:
             warnings.warn("changed metbolite annotations from {old} to {new}".format(
-                old=str(self.annotations),
-                new=str(new_annotations)))
-        self.annotations = new_annotations
+                new=str(self.annotations),
+                old=str(old_annotations)))
 
-    def add_annotations(self, new_annotations: dict) -> int:
+    def add_annotations(self, new_annotations: dict, source:str) -> int:
         """ append new annotations to the dict of metabolite annotations
         check if there have been actually added new annotation, if so return 1,else 0"""
         old_annotation = deepcopy(self.annotations)
         for x in new_annotations.keys():
             if x in self.annotations.keys():
                 self.annotations[x].extend(new_annotations[x])
                 # remove duplicates and sort list
-                new_annolst = list(set(self.annotations[x]))
-                new_annolst.sort()
-                self.annotations[x] = new_annolst
+                new_annolst = list(dict.fromkeys(self.annotations[x]))
+                new_sources = self.annotations_source[x]
+                new_sources.extend([source]*(len(new_annolst) - len(old_annotation[x])))
+                new_annolst, new_sources = zip(*sorted(zip(new_annolst,new_sources)))
+                self.annotations[x] = list(new_annolst)
+                self.annotations_source[x] = list(new_sources)
             else:
                 # remove duplicates and sort list
-                new_annolst = list(set(new_annotations[x]))
-                new_annolst.sort()
-                self.annotations[x] = new_annolst
+                new_annolst = list(dict.fromkeys(new_annotations[x]))
```

Comment:
What is happening? Is there quite a bit of duplication with the code above?

Reply:
Yes and no... maybe this could be written more concise, but it won't affect performance.